virtcontainers/qemu.go

```diff
@@ -1341,7 +1394,18 @@ func (q *qemu) hotplugVFIODevice(device *config.VFIODev, op operation) (err erro
 				err = fmt.Errorf("Incorrect VFIO device type found")
 			}
 		}
-		return err
+		if err != nil {
+			return err
+		}
+		// XXX: In some circumstances we already know some or
```

Comment:
As written, the comment confuses me more than it enlightens me, because it talks about "some circumstances" that I have no real clue about. I can _guess_ that you talk about the path above where you had a `bridge` variable, for example, but even so, that declaration is local to the scope, so you can't use it here.

Not that it would be a biggie to change, but the inconsistency between the comment and the scoping muddies the waters as what you mean with "we know" - Yes, we computed it above, but we no longer have it.

Also, I think that going back to qom allows you to get the info from qemu direcly, whereas otherwise, you would just infer it.

Reply:
The variable scope isn't really the issue - that's easily altered.  The two cases here are plugging onto a bridge and plugging into a root port.

For a bridge, we actually allocate a slot ourselves (addDeviceToBridge()) and tell qemu to use it, so we already know that.  The bridge structure itself also has an address field.... but it's not actually filled in properly, at least in our case. Sigh.

For a root port we don't set an address for the device , but it will  always be zero, because, well, it's a port, there's only one place to go.  However, the root port itself has its address allocated by qemu so Kata doesn't know it until we ask qemu.  For extra inconsistency the handling of the root ports is more or less open-coded, not using the sort-of-general-but-not-really bridge handling code.  Sigh, again.

Hence the decision to just query it out for all cases.  I very much want to clean this up, but that's a later project for Kata2.

Reply:
> The bridge structure itself also has an address field.... but it's not actually filled in properly, at least in our case. Sigh.
Can you point to where this is happening?

Also, same as @c3d, I was confused after reading the comment here. Can you add the above explanation to the comments here. As far as hotplug goes, there are 3 scenarios we deal with today - pci root bus, pci/pcie bridge and pcie root ports. I want to make sure all those scenarioes are taken into account here.